### ISSUE TITLE
Ant Builds Are Failing for FlexJS Projects

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#22]
Testing with the latest build, I got errors when trying to build a FlexJS project from Ant:
1. Open Moonshine
2. Create a new FlexJS project
3. Make sure that you have selected a FlexJS SDK
4. Project > Build with Apache Ant

I get an error like this:
    : Ant build Running
    : Buildfile: /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/build.xml
    :      [echo] SDK : /Users/joelanderson/Downloads/MoonshineSDKs/Flex_SDK/Flex_4.16.0_AIR_25.0
    :   [taskdef] Could not load definitions from resource flexTasks.tasks. It could not be found.
    : init:
    :    [delete] Deleting directory /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/DEPLOY
    :    [delete] Deleting directory /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/build-debug
    :     [mkdir] Created dir: /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/DEPLOY
    :     [mkdir] Created dir: /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/build-debug
    : compileWeb:
    : BUILD FAILED
    : /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/build.xml:76: Problem: failed to create task or type mxmlc
    : Cause: The name is undefined.
    : Action: Check the spelling.
    : Action: Check that any custom tasks/types have been declared.
    : Action: Check that any <presetdef>/<macrodef> declarations have taken place.
    : Total time: 0 seconds
    : BUILD FAILED
    : /Users/joelanderson/Documents/Flex/temp_workspace/TestANTFlexJS/build/build.xml:76: Problem: failed to create task or type mxmlc
    : Cause: The name is undefined.
    : Action: Check the spelling.
    : Action: Check that any custom tasks/types have been declared.
    : Action: Check that any <presetdef>/<macrodef> declarations have taken place.
    : Total time: 0 seconds
    : FSCH exit code: 1

Checking the FlexJS SDK, I noticed that lib/falcon-flexTasks.jar does not exist in FlexJS 0.8.0.    I found it at js/lib/falcon-flexTasks.jar instead:
    <taskdef resource="flexTasks.classpath" tasks="${FALCON_HOME}/lib/falcon-flexTasks.jar"/>

However, updating this line to reference "${FALCON_HOME}/js/lib/falcon-flexTasks.jar" gave the same error.

I tested on Mac OS X 10.10.3 with Flex JS 0.8.0, Flash 25.0.

I'm trying to set up a test on Windows 10, but I am having trouble getting a FlexJS SDK installed.

**Comment by rat-moonshine** [comment]
I noticed the Ant build file is a copy from other project template. In a fact, there was nothing to build separately through an Ant build process in case of FlexJS projects. The "Build as JavaScript" or "Build & Run as JavaScript" does the requisite production job itself - by creating both debug and production output versions in the project's "bin" folder. Thus one can easily take the "bin/js-release" version and use it per needs. Having a separate Ant build process to produce "js-release" version, I think somewhat an extra for FlexJS projects. We can probably get rid of it (somehow).

**Comment by rat-moonshine** [comment]
While testing the procedure, I noticed another problem though - the latest build somehow is trying to open the swf version of project even I chose "Build & Run as JavaScript". Checking.

**Comment by rat-moonshine** [comment]
While testing with newer version of Moonshine (from Bamboo), I couldn't able to produce the problem anymore.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 Can you help building a basic FlexJS Ant build file (the kind of Ant build script with supply with our project templates) to produce the "bin-release" folder?

I tried to modify an existing Ant build script to do that (project zip file attached, the Ant file you can find inside "build" folder), but not sure what I'm missing, or even the MXMLC jar file I pointed is right in case of FlexJS project or not.

To run the Ant build file I used this basic script:
`ant -DFALCON_HOME=c:\FlexJSSDKFolderPath`

The build terminated with unexpected error, though:
![image](https://user-images.githubusercontent.com/24370870/30805785-a6c18904-a210-11e7-8cf9-98c255470ede.png)

Can you please, take a look (?)

**Comment by JoelProminic** [comment]
The Ant scripts could also run the commands directly.  This works on Mac OS:
```
	<!-- Build the application as Javascript -->
	<target name="compileJavascript" depends="init" >
		<exec executable="${FLEX_HOME}/js/bin/mxmlc">
			<arg value="${SOURCE_APP_FILE}" />
		</exec>
	
	</target>
```

You might need to add "cmd" and "-c" options to get it to work on Windows.   Unfortunately, I have been having trouble getting an SDK to test on Windows.

**Comment by piotrzarzycki21** [comment]
@JoelProminic Thanks for the tip it was really helpful, cause when I tried to use mxmlc.jar it didn't recognize js:Application - not sure why.

@rat-moonshine I'm attaching build.xml file it should help build template for Ant build to JS only application. Replace your build.xml using mine. 
[build_flexjs_js_only_ant.zip](https://github.com/prominic/Moonshine-IDE/files/1331055/build_flexjs_js_only_ant.zip)

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 Thanks Piotr, I'm taking this from here. I shall test the Ant file soon (one I done with my running tasks) and make other requisite changes for the task procedure. 

Thank you @JoelProminic, too.

**Comment by piotrzarzycki21** [comment]
I just pushed fix where I've updated ANT template for FlexJS project which build project to JavaScript. 
I tested it on Windows 10. It's need to be tested on Mac OSX - if someone could do it for me that would be good. 

I will close the issue once we confirmed that ANT build is working on Mac OSX.

**Comment by rat-moonshine** [comment]
I'm not sure where you @piotrzarzycki21 pushed the fixes to FlexJS Ant template, or how I suppose to test the fix. Rather I tried downloading the zip file you posted earlier and tried to build a FlexJS application. 

This is the command I ran to build the Ant in command-line:
```
E:\New folder\NewFlexJSBrowserProject\build>ant -DFALCON_HOME=E:\DevareaLocal\Ap
acheFlexJSFrameworkSource\source\flex-asjs\out\apache-flex-flexjs-0.9.0-bin
Buildfile: E:\New folder\NewFlexJSBrowserProject\build\build.xml
```
The output came somewhat confusing to me, though. I'm not sure which SDK its actually using (looking through the output) when I explicitly injected the "FALCON_HOME" path. This also do not produce expected HTML files in DEPLOY folder.

![image](https://user-images.githubusercontent.com/24370870/31066625-d2b2a4f2-a76b-11e7-91a0-f8e3b5d019ec.png)

The above output is showing some mix of passing "FALCON_HOME" path and the default Flex SDK path from the environment variable.

**Comment by rat-moonshine** [comment]
The updated Ant build file is using 'compiler-targets' during build process, which is available from 0.8+ FlexJS SDK. To support the Ant file against prior version of 0.8 we need some conditional codes to add/remove the 'compiler-targets' argument. 

A new property been added to the Ant template file `IS_FLEXJS_AFTER_7` (BOOL) which set by Moonshine with appropriate value during Ant build run. The Ant file now suppose to add/remove `compiler-targets` argument based on this new property value.

**Comment by rat-moonshine** [comment]
Tried to add conditional statement inside exec command in Ant build file, but it refused to add as exec command do not permits that.

Looking through the API (https://ant.apache.org/manual/Tasks/exec.html) didn't found any probable way to add such conditional statement under exec either. 

I have added a basic conditional statement to the Ant build process now, which run the certain exec process, and this works per our needs. If we have a better way to handle this, that is good, too.
